libraries/rush-lib/src/logic/ProjectChangeAnalyzer.ts

```diff
@@ -42,8 +42,26 @@ export interface IGetChangedProjectsOptions {
    * and exclude matched files from change detection.
    */
   enableFiltering: boolean;
+
+  /**
+   * If provided, this filtering function will restrict the set of changed files, and only projects
+   * containing files that pass the filter will be included.
+   *
+   * The filtering function accepts a project as an argument, and should return a promise for a function that
+   * accepts a path as a string and returns a boolean. This "factory pattern" allows the caller to
+   * perform additional asynchronous logic (loading configuration files, etc.) when encountering a
+   * project for the first time.
+   */
+  filterByProject?: ProjectFilterFactory;
```

Comment:
`filterByProject` should be completely replacing `enableFiltering`. If you want you can provide a helper method that provides an implementation of `filterByProject` backed by `RushProjectConfiguration`, though. The intent is to make it so that ProjectChangeAnalyzer no longer contains any reference to `RushProjectConfiguration` internally.